flink-connector-influxdb2/src/main/java/org/apache/flink/streaming/connectors/influxdb/source/http/WriteAPIHandler.java

```diff
@@ -0,0 +1,116 @@
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.flink.connector.base.source.reader.synchronization.FutureCompletingBlockingQueue;
+import org.apache.flink.streaming.connectors.influxdb.common.DataPoint;
+import org.apache.flink.streaming.connectors.influxdb.common.InfluxParser;
+
+@Slf4j
+public final class WriteAPIHandler extends Handler {
+    private final InfluxParser parser = new InfluxParser();
+    private final int maximumLinesPerRequest;
+    private final FutureCompletingBlockingQueue ingestionQueue;
+    private final int threadIndex;
+    private final long enqueueWaitTime;
+
+    public WriteAPIHandler(
+            final int maximumLinesPerRequest,
+            final FutureCompletingBlockingQueue ingestionQueue,
+            final int threadIndex,
+            final long enqueueWaitTime) {
+        this.maximumLinesPerRequest = maximumLinesPerRequest;
+        this.ingestionQueue = ingestionQueue;
+        this.threadIndex = threadIndex;
+        this.enqueueWaitTime = enqueueWaitTime;
+    }
+
+    @Override
+    public void handle(final HttpExchange t) throws IOException {
+        final BufferedReader in =
+                new BufferedReader(
+                        new InputStreamReader(t.getRequestBody(), StandardCharsets.UTF_8));
+
+        try {
+            String line;
+            final List<DataPoint> points = new ArrayList<>();
+            int numberOfLinesParsed = 0;
+            while ((line = in.readLine()) != null) {
+                final DataPoint dataPoint = this.parser.parseToDataPoint(line);
+                points.add(dataPoint);
+                numberOfLinesParsed++;
+                if (numberOfLinesParsed > this.maximumLinesPerRequest) {
+                    throw new RequestTooLargeException(
+                            String.format(
+                                    "Payload too large. Maximum number of lines per request is %d.",
+                                    this.maximumLinesPerRequest));
+                }
+            }
+
+            final boolean result =
+                    CompletableFuture.supplyAsync(
+                                    () -> {
+                                        try {
+                                            return this.ingestionQueue.put(
+                                                    this.threadIndex, points);
+                                        } catch (final InterruptedException e) {
+                                            return false;
+                                        }
+                                    })
+                            .get(this.enqueueWaitTime, TimeUnit.SECONDS);
+
+            if (!result) {
+                throw new TimeoutException("Failed to enqueue");
+            }
```

Comment:
This should probably be solved in the `FutureCompletingBlockingQueue` but since we cannot extend it here, I'd leave it as is. But you are effectively using a third thread here.